federation/plugin/pkg/admission/schedulingpolicy/admission.go

```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schedulingpolicy
+
+import (
+	"fmt"
+	"io"
+	"time"
+
+	"github.com/golang/glog"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/ref"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	kubeapiserveradmission "k8s.io/kubernetes/pkg/kubeapiserver/admission"
+)
+
+const (
+	pluginName               = "SchedulingPolicy"
+	configKey                = "schedulingPolicy"
+	policyConfigMapNamespace = "kube-federation-scheduling-policy"
+	minRetryBackoff          = time.Millisecond
+	maxRetryBackoff          = time.Second * 60
+	defaultRetryBackoff      = time.Millisecond * 100
+)
+
+type admissionConfig struct {
+	Kubeconfig   string        `json:"kubeconfig"`
+	RetryBackoff time.Duration `json:"retryBackoff"`
+}
+
+type admissionController struct {
+	*admission.Handler
+	policyEngineClient       *rest.RESTClient            // client to communicate with policy engine
+	policyEngineRetryBackoff time.Duration               // backoff for policy engine queries
+	client                   internalclientset.Interface // client to communicate with federation-apiserve
+}
+
+var _ = kubeapiserveradmission.WantsInternalKubeClientSet(&admissionController{})
+
+func newAdmissionController(file io.Reader) (*admissionController, error) {
+
+	config, err := loadConfig(file)
+	if err != nil {
+		return nil, err
+	}
+
+	policyEngineClient, err := loadRestClient(config.Kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &admissionController{
+		Handler:                  admission.NewHandler(admission.Create, admission.Update),
+		policyEngineClient:       policyEngineClient,
+		policyEngineRetryBackoff: config.RetryBackoff,
+	}
+
+	return c, nil
+}
+
+func (c *admissionController) Validate() error {
+	if c.client == nil {
+		return fmt.Errorf("%s requires a client", pluginName)
+	}
+	return nil
+}
+
+func (c *admissionController) SetInternalKubeClientSet(client internalclientset.Interface) {
```

Comment:
As discussed offline with @caesarxuchao, admission controllers (like this one) also talk to apiserver directly to fetch other resources (like configmaps and events in this case).
Admission controllers should be using versioned clientsets for that.

This PR is fine as is since we do not have a way to provide versioned clients today, but adding support for it should not be too difficult. We can add another method SetVersionedKubeClientSet similar to SetInternalKubeClientSet.
This will become critical as we move admission controllers out of repo without moving the internal API structs.